### PR TITLE
ECC: DH implementation and self tests

### PR DESCRIPTION
**Pull request by pascal-brand-st-dev**
Signed-off-by: Pascal Brand <pascal.brand@st.com>

### CODE REVIEW

File: core/lib/libtomcrypt/src/tee_ltc_provider.c [hunk 1]
```diff
@@ -1491,53 +1554,96 @@ static TEE_Result ecc_verify(uint32_t algo, struct ecc_public_key *key,
...
+	ltc_res = ecc_shared_secret(&ltc_private_key, &ltc_public_key,
+				    secret, secret_len);
+	res = ((ltc_res == CRYPT_OK) ? TEE_SUCCESS : TEE_ERROR_BAD_PARAMETERS);
+
+out:
+	mp_clear_multi(key_z, NULL);
 	return res;
 }
 
+#if 0
```
Comment (jenswi-linaro) [inline core/lib/libtomcrypt/src/tee_ltc_provider.c]: Do we need this?

Comment (pascal-brand-st-dev) [inline core/lib/libtomcrypt/src/tee_ltc_provider.c]: No. I forgot to clean it. I will.

**Comment by pascal-brand-st-dev** [comment]
#if 0 code has been removed

**Comment by jforissier** [comment]
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`
`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey 32 & 64-bit)`

File: core/lib/libtomcrypt/src/tee_ltc_provider.c [hunk 2]
```diff
@@ -1491,53 +1554,76 @@ static TEE_Result ecc_verify(uint32_t algo, struct ecc_public_key *key,
...
-		return res;
+	ltc_res = mp_init_multi(&key_z, &r, &s, NULL);
+	if (ltc_res != CRYPT_OK)
+		return TEE_ERROR_OUT_OF_MEMORY;
 
-	res = ecc_compute_key_idx(&ltc_key, key_size_bytes);
-	if (res != TEE_SUCCESS)
-		return res;
+	res = ecc_populate_ltc_public_key(&ltc_key, key, key_z, algo,
+					  &key_size_bytes);
```
Comment (jenswi-linaro) [inline core/lib/libtomcrypt/src/tee_ltc_provider.c]: Not checking return value

File: core/lib/libtomcrypt/src/tee_ltc_provider.c [hunk 3]
```diff
@@ -1491,53 +1554,76 @@ static TEE_Result ecc_verify(uint32_t algo, struct ecc_public_key *key,
...
 
 	/* check keysize vs sig_len */
 	if ((key_size_bytes * 2) != sig_len)
-		return TEE_ERROR_BAD_PARAMETERS;
+		goto out;
```
Comment (jenswi-linaro) [inline core/lib/libtomcrypt/src/tee_ltc_provider.c]: What value will be returned here?

File: core/lib/libtomcrypt/src/tee_ltc_provider.c [hunk 4]
```diff
@@ -1491,53 +1554,76 @@ static TEE_Result ecc_verify(uint32_t algo, struct ecc_public_key *key,
...
+	if (res != TEE_SUCCESS)
+		goto out;
+	res = ecc_populate_ltc_public_key(&ltc_public_key, public_key, key_z,
+					  0, &key_size_bytes);
+	if (res != TEE_SUCCESS)
+		goto out;
+
+	ltc_res = ecc_shared_secret(&ltc_private_key, &ltc_public_key,
+				    secret, secret_len);
+	res = ((ltc_res == CRYPT_OK) ? TEE_SUCCESS : TEE_ERROR_BAD_PARAMETERS);
```
Comment (jenswi-linaro) [inline core/lib/libtomcrypt/src/tee_ltc_provider.c]: Minor remark: I would have preferred an ordinary if statement.

**Comment by pascal-brand-st-dev** [comment]
Review taken into account

**Comment by jenswi-linaro** [comment]
```
Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>
```

**Pull request merged**